crates/ruff_python_parser/src/lexer.rs

```diff
@@ -847,39 +999,66 @@ impl<'source> Lexer<'source> {
                 if self.cursor.eat_char('=') {
                     Tok::NotEqual
                 } else {
-                    return Err(LexicalError {
-                        error: LexicalErrorType::UnrecognizedToken { tok: '!' },
-                        location: self.token_start(),
-                    });
+                    Tok::Exclamation
                 }
             }
             '~' => Tok::Tilde,
             '(' => {
+                if let Some(fstring_context) = self.fstring_stack.last_mut() {
+                    fstring_context.open_parentheses();
+                }
                 self.nesting += 1;
```

Comment:
It is unfortunate that we now need to increment open parentheses twice if inside of an `fstring` context (and we have more branching inside a hot function. 

Would it be possible to store the initial `parentheses_count` on the `FStringContext` instead? You can then compute the fstring parentheses by doing `self.nesting.saturating_sub(self.initial_count)`

Reply:
This is a good point! This would mean that to get the `parentheses_count` inside any method of `FStringContext`, the `self.nesting` needs to be passed to these methods.

Also, note that for closing parentheses we still need to call the method to decrement the format spec count.

Reply:
I've made this change. It turns out that the `last_mut` part was taking on average 10% of the lexing time. I think this is because we were performing this check for _every_ parentheses irrespective of whether we're inside a f-string or not. This has improved the benchmarks. Thanks!